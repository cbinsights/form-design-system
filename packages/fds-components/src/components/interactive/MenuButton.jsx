import React from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';
import Flex from '../layout/Flex';
import FlexItem from '../layout/FlexItem';

/**
 * @module MenuButton
 * @param {Object} props react props
 * @returns {ReactElement}
 */
const MenuButton = (props) => {
  const {
    iconComponent,
    label,
    className,
    labelClassName,
    disabled,
    active,
    ...other
  } = props;

  return (
    <Button
      className={cx(
        'menuButton',
        {
          'menuButton--active': active,
          'menuButton--disabled': disabled,
        },
        className
      )}
      disabled={disabled}
      variant="text"
      {...other}
    >
      <Flex className="menuButton-inner" direction="column" noGutters>
        <FlexItem className="menuButton-icon">{iconComponent}</FlexItem>
        <FlexItem shrink className={cx('typemod--tiny menuButton-label', labelClassName)}>
          {label}
        </FlexItem>
      </Flex>
    </Button>
  );
};

MenuButton.propTypes = {
  /** Icon component to render in the button */
  iconComponent: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node])
    .isRequired,

  /** Button label */
  label: PropTypes.string.isRequired,

  /** Disabled mode */
  disabled: PropTypes.bool,

  /** Active mode */
  active: PropTypes.bool,

  /** Class for label element */
  labelClassName: PropTypes.string,

  /** Optional className for inner Button component */
  className: PropTypes.string,
};

export default MenuButton;
