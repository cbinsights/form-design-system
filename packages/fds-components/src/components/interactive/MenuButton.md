
### MenuButton

Navigation button that renders an icon above a small label. Uses `material-ui`
button as a base. These buttons can be used for either navigation or user actions.

This commponent spreads `...other` props on the `Button` component. This allows
passing of props defined in the [`material-ui` `Button` API](https://material-ui.com/api/button/#props).

**The `iconComponent` prop accepts any arbitrary JSX**. Use this to render an
SVG icon, an icon font, or whatever you need.

Additional styling customization should be done through `className` and
`labelClassName` props.

#### Basic MenuButton

```js
<MenuButton
  iconComponent={<span className="typemod--xlarge">🔥</span>}
  label="Set on fire"
  onClick={() => { alert('clicked') }}
/>
<MenuButton
  iconComponent={<span className="typemod--xlarge">🍕</span>}
  label="Long pizza button label"
/>
```

#### Disabled/Active MenuButtons

```js
<MenuButton
  iconComponent={<span className="typemod--xlarge">📞</span>}
  label="Call"
  disabled
/>
<MenuButton
  iconComponent={<span className="typemod--xlarge">✏️</span>}
  label="Edit Mode"
  active
/>
```
