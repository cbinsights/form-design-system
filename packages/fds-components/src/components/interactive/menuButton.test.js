import React from 'react';
import { shallow } from 'enzyme';

import MenuButton from './MenuButton';

const renderComponent = (props) => shallow(<MenuButton {...props} />);

describe('MenuButton component', () => {

  it('matches snapshot (default props)', () => {
    const component = renderComponent({
      label: 'My Label',
      iconComponent: (<span>ðŸŒ­</span>),
    });
    expect(component).toMatchSnapshot();
  });

  it('matches snapshot (set all props)', () => {
    const component = renderComponent({
      disabled: true,
      onClick: () => false,
      labelClassName: 'myCustomClass',
      className: 'foo',
      label: 'My Label',
      iconComponent: (<span>ðŸŒ­</span>),
    });
    expect(component).toMatchSnapshot();
  });
});
