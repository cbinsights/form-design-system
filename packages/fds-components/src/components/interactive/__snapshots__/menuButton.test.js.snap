// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuButton component matches snapshot (default props) 1`] = `
<WithStyles(Button)
  className="menuButton"
  variant="text"
>
  <Flex
    align="stretch"
    className="menuButton-inner"
    direction="column"
    noGutters={true}
  >
    <FlexItem
      className="menuButton-icon"
      shrink={false}
    >
      <span>
        ðŸŒ­
      </span>
    </FlexItem>
    <FlexItem
      className="typemod--tiny menuButton-label"
      shrink={true}
    >
      My Label
    </FlexItem>
  </Flex>
</WithStyles(Button)>
`;

exports[`MenuButton component matches snapshot (set all props) 1`] = `
<WithStyles(Button)
  className="menuButton menuButton--disabled foo"
  disabled={true}
  onClick={[Function]}
  variant="text"
>
  <Flex
    align="stretch"
    className="menuButton-inner"
    direction="column"
    noGutters={true}
  >
    <FlexItem
      className="menuButton-icon"
      shrink={false}
    >
      <span>
        ðŸŒ­
      </span>
    </FlexItem>
    <FlexItem
      className="typemod--tiny menuButton-label myCustomClass"
      shrink={true}
    >
      My Label
    </FlexItem>
  </Flex>
</WithStyles(Button)>
`;
